src/include/duckdb/common/file_system.hpp

```diff
@@ -36,6 +36,40 @@ class Logger;
 class ClientContext;
 class QueryContext;
 
+template <typename T>
+struct PaginatedResult {
+	PaginatedResult() = default;
+
+	virtual ~PaginatedResult() = default;
+
+	bool FetchNextPage() {
+		if (page_buffer.empty()) {
+			page_buffer = GetNextPageInternal();
+		}
+		return !page_buffer.empty();
+	}
+
+	vector<T> ClaimNextPage() {
+		return std::move(page_buffer);
+	}
+
+	vector<T> ClaimAllPages() {
+		vector<T> result;
+		while (FetchNextPage()) {
+			auto next_page = ClaimNextPage();
+			result.insert(result.end(), std::make_move_iterator(next_page.begin()),
+			              std::make_move_iterator(next_page.end()));
+		}
+		return result;
+	}
+
+protected:
+	virtual vector<T> GetNextPageInternal() = 0;
+
+private:
+	std::vector<T> page_buffer;
```

Comment:
nit: should we use `duckdb::vector` instead of STL one, which provides out-of-bound check?